src/lib/OccAnalysis.hs

```diff
@@ -221,16 +220,23 @@ summary atom = case atom of
 unknown :: HoistableE e => e n -> OCCM n (IxExpr n)
 unknown obj = do
   freeIxs <- mapM ixExpr $ freeAtomVarsList obj
-  let freeIterationNames = nub $ concatMap freeAtomVarsList freeIxs
-  return $ Deterministic $ freeIterationNames
+  if any (== IxAll) freeIxs then
+    return IxAll
+  else
+    let freeIterationNames = nub $ concatMap freeAtomVarsList freeIxs
+    -- TODO(correctness) Should we detect whether `obj` is an expression with a
+    -- funny effect like `IO` and fall back from `Deterministic` to `IxAll`?
+    -- (For reader, writer, and state effects such a fallback should not be
+    -- necessary, because the summary of the reference should already have any
+    -- necessary `IxAll` in it.)
+    in return $ Deterministic $ freeIterationNames
```

Comment:
Right, can't we always return IxAll in that case? Since this is still unsound.

Reply:
Done.